server/src/main/java/io/crate/types/TypeCompatibility.java

```diff
@@ -100,15 +100,11 @@ private static DataType<?> typeCompatibilityForParametrizedType(DataType<?> from
                 }
             }
             return null;
-        } else {
-            if ((fromType.id() == StringType.ID && toType.id() == StringType.ID)
-                || (fromType.id() == CharacterType.ID && toType.id() == CharacterType.ID)) {
-                if (((StringType) fromType).lengthLimit() > ((StringType) toType).lengthLimit()) {
-                    return fromType;
-                } else {
-                    return toType;
-                }
+        } else if (fromType.id() == toType.id() && fromType.characterMaximumLength() != null && fromType.characterMaximumLength().equals(toType.characterMaximumLength())) {
```

Comment:
not using `Object.equals `since should not be entering this when both are maxLens are `null`